### PR TITLE
[openshift-upgrade-watcher] introduce new integration

### PR DESCRIPTION
**Pull request by maorfr**
part of https://issues.redhat.com/browse/APPSRE-2514

this integration watches for the UpgradeConfig resource in each cluster and notifies in a slack channel if the cluster is currently being upgraded.

### CODE REVIEW

File: reconcile/openshift_upgrade_watcher.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+            if state.exists(state_key):
+                # already notified
+                continue
+            logging.info(['cluster_upgrade', cluster])
+            if not dry_run:
+                state.add(state_key)
+                slack.chat_post_message(
+                    f'Heads up! cluster `{cluster}` is currently ' +
+                    f'being upgraded to version `{version}`'
+                )
```
Comment (apahim) [inline reconcile/openshift_upgrade_watcher.py:66]: I love it!
Can we somehow detect when it finishes?

File: reconcile/openshift_upgrade_watcher.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        state_key = f'{cluster}-{upgrade_at}'
+        # if this is the first iteration in which 'now' had passed
+        # the upgrade at date time, we send a notification
+        if upgrade_at_obj < now:
+            if state.exists(state_key):
+                # already notified
+                continue
+            logging.info(['cluster_upgrade', cluster])
+            if not dry_run:
+                state.add(state_key)
```
Comment (rrati) [inline reconcile/openshift_upgrade_watcher.py:62]: DO you need to remove this key from the state file at some point or at these keys unique per upgrade?

File: reconcile/openshift_upgrade_watcher.py (hunk 1 above)
Comment (maorfr) [reply reconcile/openshift_upgrade_watcher.py:66]: we currently detect it and a MR is submitted to app-interface.
we can totally enhance this later on to also report the finish. would you agree to this in a next iteration?

File: reconcile/openshift_upgrade_watcher.py (hunk 2 above)
Comment (maorfr) [reply reconcile/openshift_upgrade_watcher.py:62]: the key contains a timestamp, so it's unique per upgrade

File: reconcile/openshift_upgrade_watcher.py (hunk 1 above)
Comment (maorfr) [reply reconcile/openshift_upgrade_watcher.py:66]: although even this MR does not indicate the real finish. nor does the `ClusterVersion` on the cluster. so maybe a bit more work to report the "real" finish time.

**Review by apahim — APPROVED** [review]

**Pull request merged**